### PR TITLE
scripts: Add push_pkg.sh to enhance package development workflow (v2)

### PR DESCRIPTION
**Pull request by lemoer**
v2: In contrast to the last patches, this is now built on top of ssh
    only, without using e.g. 9pfs. Furthermore it works also with
    arbitary remote hosts on any target/architecture. Also the
    scripts were renamed and moved to /scripts.

The aim of this commit is to allow fast rebuild cycles during the
development of gluon packages.

Currently the following workflow can be used:

	# start a local qemu instance
	scripts/run_qemu.sh output/images/factory/[...].img

	# do your changes in the file you want to patch
	vi package/gluon-ebtables/files/etc/init.d/gluon-ebtables

	# rebuild and update the package
	scripts/push_pkg.sh package/gluon-ebtables/

	# test your changes
	...

	# do more changes
	...

	# rebuild and update the package
	scripts/push_pkg.sh package/gluon-ebtables/

	# test your changes
	...

	(and so on...)

Implementation details:
- Currently this is based on ssh/scp.
- Opkg is used to install/update the packages in the remote machine.

Benefits:
- This works with compiled and non-compiled packages.
- This works with native OpenWrt and Gluon packages.
- This even performs the check_site.lua checks as they are integrated
  as post_install scripts into the openwrt package.
- It works for all architectures/targets.

EDIT:
- 2020-12-20: I updated this comment to the description of the new patch.

### CODE REVIEW

File: contrib/start_qemu.sh
Comment (mweinelt) [inline contrib/start_qemu.sh:8]:
```
-s              shorthand for -gdb tcp::1234
```

Comment (mweinelt) [reply contrib/start_qemu.sh:8]: FWIW: I'm opting for better readability here.

File: contrib/start_qemu.sh
Comment (mweinelt) [inline contrib/start_qemu.sh:14]: This works well for the config mode, but not for when the normal runmode.

Comment (lemoer) [reply contrib/start_qemu.sh:14]: Of course. This could be resolved later. :-)

**Comment by lemoer** [comment]
I would rather like to discuss this from a conceptual point of view first. If we decide that we want to have such a feature and which technologies should be used, we can go into detail later.

**Comment by lemoer** [comment]
I updated the patch to remove the ugly `9pfs` stuff and completely rely on ssh. Please review and comment.

**Comment by lemoer** [comment]
This commit significantly reduces iteration time for package changes:
``` shell
# time sh scripts/push_pkg.sh package/gluon-mesh-vpn-wireguard
...

________________________________________________________
Executed in    1.57 secs   fish           external 
   usr time  740.34 millis  1032.00 micros  739.31 millis 
   sys time  336.54 millis  197.00 micros  336.34 millis 
```
This time includes rebuilding the package and deploying it.

**Comment by AiyionPrime** [comment]
Just tested, works very well.

**Comment by NeoRaider** [comment]
I think it would make sense to put this in the `contrib` directory, where other scripts that aren't directly involved in Gluon's build process live.

File: scripts/push_pkg.sh
Comment (NeoRaider) [inline scripts/push_pkg.sh:78]: A better way find out what packages a Makefile contains would be:
```sh
make TOPDIR="${topdir}" -C "${pkgdir}" DUMP=1 | awk '/^Package: / { print $2 }'
```
The DUMP feature is also what OpenWrt uses internally for enumerate packages.

File: scripts/push_pkg.sh
Comment (NeoRaider) [inline scripts/push_pkg.sh:117]: How about putting all these commands into a single 'ssh' run? Would just need a `set -e` to abort on the first error.

File: package/gluon.mk
Comment (NeoRaider) [inline package/gluon.mk:39]: Hmm, I'm don't like this change... I'll see if I can come up with an alternative solution.

File: scripts/push_pkg.sh
Comment (NeoRaider) [inline scripts/push_pkg.sh:31]: The behaviour of the option handling is somewhat confusing... If `-r` is used after `-p`, the `-p` will be ignored.

How about selecting the port after all options have been parsed, using the following logic?

- If `-p` was given, always use that port
- Else-if `-r` was given, default to port 22
- Else, default to port 2223

File: scripts/push_pkg.sh
Comment (NeoRaider) [inline scripts/push_pkg.sh:44]: Use `grep -q` instead of `grep >/dev/null`

File: scripts/run_qemu.sh
Comment (NeoRaider) [inline scripts/run_qemu.sh:11]: In my qemu script I also have the following lines:
```
        -chardev stdio,id=char0,signal=off \
        -serial chardev:char0 \
```
This connects the serial console of the VM to the terminal running the script. `signal=off` might be controversial - it makes `Ctrl-C` etc. work as expected in the serial console instead of terminating qemu, but in consequence you need a second terminal if you actually want to kill qemu.

Comment (lemoer) [reply scripts/run_qemu.sh:11]: The proper way to exit qemu with `signal=off` is to press `Ctrl-a`, `Ctrl-c` and type `quit`.

Comment (NeoRaider) [reply scripts/run_qemu.sh:11]: Good to know, I wasn't aware qemu had an escape key like that.

Comment (lemoer) [reply scripts/run_qemu.sh]: I added a note to `run_qemu.sh`, so hopefully other users will stumble across that hint aswell.

File: package/gluon.mk
Comment (lemoer) [reply package/gluon.mk:39]: Please let me know when you have decided what to do here.

Or maybe you could clarify why you don't like the approach? Then I could think about a solution too.

**Comment by mweinelt** [comment]
The walkthrough in the initial post sounds good. It would be good to have documented.

File: package/gluon.mk
Comment (mweinelt) [reply package/gluon.mk:39]: It has been a month. Is this something that we can address another time or does it prevent merging this?

**Comment by lemoer** [comment]
@mweinelt done.

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:25]:
```suggestion
  contrib/run_qemu.sh output/images/factory/[...]-x86-64.img
```

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:27]:
```suggestion
  # apply changes to the desired package
```

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:39]:
```suggestion
  # rebuild and push the package to the running vm
```

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:53]:
```suggestion
* File transfers use the scp utility.
```

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:54]:
```suggestion
* Packages are installed/updated using ``opkg`` on the virtual machine.
```

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:56]:
```suggestion
* It will not override ``/etc/config`` with package defaults by default. (See the cli switch ``-P``).
```

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:63]: Except `run_qemu.sh` uses KVM on x86_64. The relationship between `run_qemu.sh` and developing on actual hardware isn't clear from this.

File: docs/dev/packages.rst
Comment (mweinelt) [inline docs/dev/packages.rst:71]:
```suggestion
* ``push_pkg.sh`` can also be used to push packages to other targets than your native arch.
```

Is this the intended meaning?

File: contrib/run_qemu.sh
Comment (mweinelt) [inline contrib/run_qemu.sh:9]:
```
-s              shorthand for -gdb tcp::1234
```

No harm in using the longer option, since it is much clearer what it does.

File: contrib/push_pkg.sh
Comment (mweinelt) [inline contrib/push_pkg.sh:18]:
```suggestion
	echo " -b          build only, do not push"
```

File: contrib/push_pkg.sh
Comment (mweinelt) [inline contrib/push_pkg.sh:48]: What is the idea behind the OPENWRT_ scope here? I'd say `REMOTE_BOARD` and `REMOTE_ARCH` makes more sense, but I think we'd be fine without the qualifier.

**Review by mweinelt — CHANGES_REQUESTED** [review]
Looks pretty good, a few remarks.

**Comment by lemoer** [comment]
@mweinelt Thanks, I implemented your suggestions.

File: package/gluon.mk
Comment (mweinelt) [reply package/gluon.mk:40]: And two more weeks went by. :disappointed:

Comment (AiyionPrime) [reply package/gluon.mk:40]: @neoraider As this becomes a rather stuck feedback-loop; how about meeting up in mumble for a few minutes tomorrow evening? Maybe that makes it easier to brainstorm and clearer for @lemoer what benefits the alternative solution should provide.

Comment (NeoRaider) [reply package/gluon.mk:40]: Sorry for the delay.

The issue with this piece of code is that `GluonCheckSite` is put into the ipk packages as a postinst script; by conditionally changing the definition, you are also changing the package files (which may not be rebuilt when a regular build is run later?).

Optimally, we wouldn't need different logic in `GluonCheckSite` for regular and `push_pkg.sh` builds at all. If it can't be avoided, the condition must be checked at package install time, not build time.

Comment (lemoer) [reply package/gluon.mk]: Thanks for the feedback.

> Optimally, we wouldn't need different logic in GluonCheckSite for regular and push_pkg.sh builds at all.

I didn't find a way to do this.

> If it can't be avoided, the condition must be checked at package install time, not build time.

I implemented this one.

File: scripts/check_site.lua
Comment (NeoRaider) [inline scripts/check_site.lua:1]: I'm currently evaluating whether we can switch to `lua-jsonc` unconditionally, removing the need to support two different loaders in `check_site.lua`.

Comment (NeoRaider) [reply scripts/check_site.lua:1]: https://github.com/freifunk-gluon/gluon/pull/2234

File: package/gluon.mk
Comment (NeoRaider) [reply package/gluon.mk]: > > Optimally, we wouldn't need different logic in GluonCheckSite for regular and push_pkg.sh builds at all.
> 
> I didn't find a way to do this.

I *think* I have an idea how to achieve this, but let's get https://github.com/freifunk-gluon/gluon/pull/2234 merged first.

Comment (NeoRaider) [reply package/gluon.mk]: I have implemented this in #2237.

**Comment by lemoer** [comment]
This PR is now implicitly also "level-2 blocked" by #2271, because #2271 is now blocking/influencing #2237. I would really like to get this merged soon.

**Comment by NeoRaider** [comment]
#2237 is merged, please rebase!

**Comment by lemoer** [comment]
@NeoRaider Done.

I also squashed it.

File: contrib/push_pkg.sh
Comment (NeoRaider) [inline contrib/push_pkg.sh:49]:
```suggestion
	REMOTE_OPENWRT_BOARD="$(ssh -p "${ssh_port}" "root@${ssh_host}" '. /etc/os-release; echo "$OPENWRT_BOARD"')"
```
SSH already runs the passed command through a shell.

It would also be nice to merge both SSH commands into one and split the resulting string afterwards to improve performance.

File: contrib/push_pkg.sh
Comment (NeoRaider) [inline contrib/push_pkg.sh:125]:
```suggestion
				if [ '$preserve_config' -eq 1 ]; then
```
Use single quotes for the remote commands whenever you don't need variables from the remote side interpreted.

File: contrib/push_pkg.sh
Comment (lemoer) [reply contrib/push_pkg.sh]: Done.

File: contrib/push_pkg.sh
Comment (lemoer) [reply contrib/push_pkg.sh]: Done.

**Comment by lemoer** [comment]
@NeoRaider Updated once again.

File: contrib/push_pkg.sh
Comment (NeoRaider) [inline contrib/push_pkg.sh:147]: There are a lot more `\"` in this section that could be replaced with `'`.

Comment (lemoer) [reply contrib/push_pkg.sh:147]: Done.

**Comment by lemoer** [comment]
@NeoRaider And updated once again.

File: contrib/push_pkg.sh
Comment (NeoRaider) [inline contrib/push_pkg.sh:145]: Hmm, thinking about this section again, you have actually found a bug in Gluon's package Makefiles!

opkg should never overwrite config files on package upgrades. The only reason the `preserve_config` flag is even necessary is that Gluon doesn't correctly specify `conffiles` in its Makefiles.

I'll prepare a PR to fix our Makefiles, and then we can get rid of another few lines of this script :)

Comment (NeoRaider) [reply contrib/push_pkg.sh:145]: #2276

You can keep the `preserve_config` flag, but it should map to the `--force-maintainer` option of opkg instead of handling config files manually.

Comment (lemoer) [reply contrib/push_pkg.sh]: Implemented.

**Comment by lemoer** [comment]
I implemented several changes now. Let's just freeze this now and merge it once #2276 is merged. If new ideas come, we can add them later in new PRs or directly on the master branch.

**Pull request merged**
